src/sql/src/func.rs

```diff
@@ -587,52 +584,49 @@ impl<'a> ArgImplementationMatcher<'a> {
         f
     }
 
-    /// Plans `args` as `ScalarExprs` of that match the `ParamList`'s specified types.
-    fn generate_param_exprs(
-        &self,
-        args: &[sql_parser::ast::Expr],
-        params: ParamList,
-    ) -> Result<Vec<ScalarExpr>, failure::Error> {
-        match params {
-            ParamList::Exact(p) => {
-                let mut exprs = Vec::new();
-                for (arg, param) in args.iter().zip(p.iter()) {
-                    exprs.push(self.coerce_arg_to_type(arg, param)?);
-                }
-                Ok(exprs)
-            }
-            ParamList::Repeat(p) => {
-                let mut exprs = Vec::new();
-                for (i, arg) in args.iter().enumerate() {
-                    exprs.push(self.coerce_arg_to_type(arg, &p[i % p.len()])?);
-                }
-                Ok(exprs)
-            }
-        }
-    }
-
     /// Generates `ScalarExpr` necessary to coerce `Expr` into the `ScalarType`
     /// corresponding to `ParameterType`; errors if not possible. This can only
     /// work within the `func` module because it relies on `ParameterType`.
     fn coerce_arg_to_type(
         &self,
-        arg: &Expr,
+        arg: CoercibleScalarExpr,
         typ: &ParamType,
     ) -> Result<ScalarExpr, failure::Error> {
+        /// XXX NOTE(benesch): this function and plan_coerce do not have the
+        /// right separation.
         use super::query::CastContext::*;
         let s: ScalarType = typ.into();
-        let hinted_expr = super::query::plan_expr(self.ecx, &arg, Some(s.clone()))?;
 
-        if self.ecx.scalar_type(&hinted_expr) == s {
-            Ok(hinted_expr)
-        } else if ParamType::JsonbAny == *typ {
-            super::query::plan_to_jsonb(self.ecx, self.ident, hinted_expr)
+        if let CoercibleScalarExpr::Coerced(arg) = &arg {
+            if self.ecx.scalar_type(arg) == s {
+                return Ok(arg.clone());
+            }
+        }
+
+        if ParamType::JsonbAny == *typ {
```

Comment:
This could be a `match typ` now that we're no longer comparig `arg.is_string_literal()`

Reply:
Sure thing, done.